### PR TITLE
Refactors gun icon/item code

### PR DESCRIPTION
**Pull request by Lumipharon**
## About The Pull Request
This PR refactors the procs relating to gun icon and item states, bringing them all under the parent gun item.
That means there are no more specific procs under eguns, and they should now be under the correct procs, and call to parent correctly etc.

- Added support for overlays to show ammo level remaining. This supercedes ammo level icon_states for eguns, which allowed me to remove a huge number of unnecessary sprites (like 16 for TX73 alone, jesus). This is also available for all guns, not just eguns.

- Fixed some issues with magazine overlays, Uzi extended mag is no longer blue at the end, (and it's sprite doesn't move up and down depending on if it's loaded), and the PPSh now actually has a visual difference for it's two mag types.
image

- Fixed item state updates (currently only used by certain eguns), so they actually update as you fire.
image

- Fixed SOME of the mob sprites for eguns. Honestly gun mob sprites are so fucked, it would take ages to fix them all. I did most of the eguns, removed item_state updates from pulse rifles (they recharge and have huge capacity, so is extra pointless), then had to stop before I offed myself.

I was also going to make item_state updates use overlays, and make all the TE guns use it but that seems a bit more tricky, so I'll leave that to another day and another PR.

## Changelog
:cl:
fix: Whole bunch of minor sprite fixed for eguns
imageadd: PPSh now has a separate mob sprites for its drum and sickle magazines
refactor: Refactor gun icon and item sprite code
/:cl:

### CODE REVIEW

**Review by Kuro020 — APPROVED** [review]

**Comment by Lumipharon** [comment]
Merge conflict go home.

File: code/modules/projectiles/gun_system.dm [hunk 1]
```diff
@@ -456,24 +458,54 @@
...
 	attachment_overlays[ATTACHMENT_SLOT_MAGAZINE] = overlay
-	overlays += overlay
+	. += overlay
 
 /obj/item/weapon/gun/update_item_state()
-	item_state = "[base_gun_icon][flags_item & WIELDED ? "_w" : ""]"
+	if(ammo_level_item == FALSE)
+		item_state = "[base_gun_icon][flags_item & WIELDED ? "_w" : ""]"
+	else
+		. = item_state
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_system.dm:500]: early return
also on a sidenote why . here?

File: code/modules/projectiles/gun_system.dm [hunk 2]
```diff
@@ -456,24 +458,54 @@
...
-	overlay = image(current_mag.icon, src, get_magazine_overlay(current_mag))
+	overlay = get_magazine_overlay(current_mag)
 	attachment_overlays[ATTACHMENT_SLOT_MAGAZINE] = overlay
-	overlays += overlay
+	. += overlay
 
 /obj/item/weapon/gun/update_item_state()
-	item_state = "[base_gun_icon][flags_item & WIELDED ? "_w" : ""]"
+	if(ammo_level_item == FALSE)
+		item_state = "[base_gun_icon][flags_item & WIELDED ? "_w" : ""]"
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_system.dm:498]: use ! not == false

File: code/modules/projectiles/gun_system.dm [hunk 3]
```diff
@@ -247,6 +247,19 @@
 	///what ammo to use for overcharge
 	var/ammo_diff = null
 
+/*
+ *  extra icon and item states or overlays
+*/
+
+	///Whether the gun has item_state sprites that show ammo level remaining
+	var/ammo_level_item = FALSE
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_system.dm:255]: why is this not a flag

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun_system.dm (hunk 1 above)
Comment (Lumipharon) [reply code/modules/projectiles/gun_system.dm:500]: It was like that in the previous proc under eguns so I kept it. I've changed it to it's own var.

File: code/modules/projectiles/gun_system.dm (hunk 3 above)
Comment (Lumipharon) [reply code/modules/projectiles/gun_system.dm:255]: done.

File: code/modules/projectiles/gun_system.dm (hunk 2 above)
Comment (Lumipharon) [reply code/modules/projectiles/gun_system.dm:498]: done.

File: code/modules/projectiles/gun_system.dm (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lumipharon) [reply code/modules/projectiles/gun_system.dm:255]: done.

File: code/modules/projectiles/gun_system.dm [hunk 4]
```diff
@@ -247,6 +247,19 @@
...
+	///Whether the gun has ammo level overlays for its icon, mainly for eguns
+	var/ammo_level_icon = null
```
[the remark this answers is not in the document]
Comment (Lumipharon) [reply code/modules/projectiles/gun_system.dm:257]: changed to just "var/ammo_level_icon"

File: code/modules/projectiles/gun_system.dm [hunk 5]
```diff
@@ -456,24 +458,54 @@
...
+	overlay = get_magazine_overlay(current_mag)
 	attachment_overlays[ATTACHMENT_SLOT_MAGAZINE] = overlay
-	overlays += overlay
+	. += overlay
 
 /obj/item/weapon/gun/update_item_state()
-	item_state = "[base_gun_icon][flags_item & WIELDED ? "_w" : ""]"
+	if(ammo_level_item == FALSE)
+		item_state = "[base_gun_icon][flags_item & WIELDED ? "_w" : ""]"
+	else
```
[the remark this answers is not in the document]
Comment (Lumipharon) [reply code/modules/projectiles/gun_system.dm:499]: done

File: code/modules/projectiles/gun_system.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (Lumipharon) [reply code/modules/projectiles/gun_system.dm:500]: I unno, I kept what was there previously, I think it was just being used as a var as the return wasn't being used anywhere. I changed it to its own var.

File: code/modules/projectiles/gun_system.dm [hunk 6]
```diff
@@ -247,6 +247,19 @@
...
+*/
+
+	///Whether the gun has item_state sprites that show ammo level remaining
+	var/ammo_level_item = FALSE
+	///Whether the gun has ammo level overlays for its icon, mainly for eguns
+	var/ammo_level_icon = null
+	///Whether the icon_state overlay is offset in the x axis
+	var/icon_overlay_x_offset = null
+	///Whether the icon_state overlay is offset in the Y axis
+	var/icon_overlay_y_offset = null
```
[the remark this answers is not in the document]
Comment (Lumipharon) [reply code/modules/projectiles/gun_system.dm:259]: Changed to 0. I'm guessing that's what you're saying it should be?

File: code/modules/projectiles/guns/energy.dm [hunk 7]
```diff
@@ -407,9 +334,58 @@
 
 	in_chamber = get_ammo_object(chamber_items[current_chamber_position])
 
-/obj/item/weapon/gun/energy/lasgun/lasrifle/update_item_state(mob/user) //Without this override icon states for wielded guns won't show. because lasgun overrides and this has no charge icons
-	item_state = "[initial(icon_state)][flags_item & WIELDED ? "_w" : ""]"
+//Tesla gun
+/obj/item/weapon/gun/energy/lasgun/lasrifle/tesla
+	name = "\improper Terra Experimental tesla shock rifle"
+	desc = "A Terra Experimental energy rifle that fires balls of elecricity that shock all those near them, it is meant to drain the plasma of unidentified creatures from within, limiting their abilities. Unlike the other TE Laser weapons, lasers don't come out of this weird weapon. As with all TE Laser weapons, they use a lightweight alloy combined without the need for bullets any longer decreases their weight and aiming speed quite some vs their ballistic counterparts. Uses standard Terra Experimental  (abbreviated as TE) power cells."
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/energy.dm:340]:
```suggestion
	desc = "A Terra Experimental energy rifle that fires balls of elecricity that shock all those near them, it is meant to drain the plasma of unidentified creatures from within, limiting their abilities. As with all TE Laser weapons, they use a lightweight alloy combined without the need for bullets any longer decreases their weight and aiming speed quite some vs their ballistic counterparts. Uses standard Terra Experimental (TE) power cells."
```
it's as if `A Terra Experimental energy rifle that fires balls of elecricity that shock all those near them` isn't enough

File: code/__DEFINES/conflict.dm [hunk 8]
```diff
@@ -77,6 +77,7 @@
 #define GUN_AMMO_COUNT_BY_PERCENTAGE (1<<11)
 #define GUN_AMMO_COUNT_BY_SHOTS_REMAINING (1<<12)
 #define GUN_NO_PITCH_SHIFT_NEAR_EMPTY (1<<13)
+#define ITEM_SHOWS_AMMO_REMAINING (1<<14)
```
Comment (TiviPlus) [inline code/__DEFINES/conflict.dm:81]: not an item flag but named item

**Review by TiviPlus — CHANGES_REQUESTED** [review]
rest good enough

File: code/__DEFINES/conflict.dm (hunk 8 above)
Comment (Lumipharon) [reply code/__DEFINES/conflict.dm:81]: What's a better way to name it? I said item because it's the item_state that it effects.

Comment (Lumipharon) [reply code/__DEFINES/conflict.dm]: updated

**Pull request merged**